StardewArchipelago/Locations/CodeInjections/Modded/SVE/SVEShopStockModifier.cs

```diff
@@ -0,0 +1,218 @@
+                }
+                else if (item.Id.Equals(ModItemIds.TEMPERED_HAMMER, StringComparison.InvariantCultureIgnoreCase))
+                {
+                    var apShopItem = CreateArchipelagoLocation(item, hammerLocationName);
+                    shopData.Items.RemoveAt(i);
+                    shopData.Items.Insert(i, apShopItem);
+                }
+            }
+        }
+
+        private void MakeBearBarter(ShopData shopData)
+        {
+            var berryItems = _stardewItemManager.GetObjectsWithPhrase("berry");
+            var newItems = new List<ShopItemData>();
+            foreach (var item in shopData.Items)
+            {
+                var random = new Random((int)Game1.stats.DaysPlayed + (int)Game1.uniqueIDForThisGame / 2 + item.GetHashCode());
+                var chosenItemGroup = berryItems.FindAll(x => !(x.Name.Contains("Joja") || x.Name.Contains("Seeds")) && x.SellPrice != 0 );
+                var chosenItem = chosenItemGroup[random.Next(chosenItemGroup.Count)];
+                var isRecipe = item.ItemId.Contains("Baked Berry Oatmeal") || item.ItemId.Contains("Flower Cookie");
+                if (isRecipe && _archipelago.SlotData.Chefsanity.HasFlag(Chefsanity.Purchases))
+                {
+                    chosenItem = BerryIfChefsanityIsOn();
+                }
+                newItems.Add(SwapItemToBerryBarter(item, chosenItem));
+            }
+            shopData.Items = newItems;
+        }
+
+        private StardewObject BerryIfChefsanityIsOn()
+        {
+            if (Game1.season.HasFlag(Season.Spring))
+            {
+                return _stardewItemManager.GetObjectById(SALMONBERRY_ID);
+            }
+            else if (Game1.season.HasFlag(Season.Summer))
+            {
+                return _stardewItemManager.GetObjectById(SPICEBERRY_ID);
+            }
+            else if (Game1.season.HasFlag(Season.Fall))
+            {
+                return _stardewItemManager.GetObjectById(BLACKBERRY_ID);
+            }
+            return _stardewItemManager.GetObjectById(CRYSTALFRUIT_ID);
+        }
+
+        protected ShopItemData SwapItemToBerryBarter(ShopItemData item, StardewObject berryItem)
+        {
+            var hasKnowledge = _archipelago.HasReceivedItem("Bear Knowledge");
+            var bearShopItem = item.DeepClone();
+            bearShopItem.Id = $"{ModEntry.Instance.ModManifest.UniqueID}_{berryItem.Name.Replace(" ", "_")}";
+            var divisorBonus = hasKnowledge ? 10 : 5;
+            var chosenItemExchangeRate = ExchangeRate(item.Price / divisorBonus, berryItem.SellPrice);
+            bearShopItem.TradeItemId = berryItem.Id;
+            bearShopItem.Price = 0;
+            bearShopItem.MinStack = chosenItemExchangeRate[0];
+            bearShopItem.TradeItemAmount = chosenItemExchangeRate[1];
+            return bearShopItem;
+        }
+
+    public int[] ExchangeRate(int soldItemValue, int requestedItemValue)
```

Comment:
Is this the exact same Exchange Rate algorithm as in the Junimo shops? Why not centralize it then?

Reply:
I would recommend an extra layer of polymorphism between both of the ShopStockModifiers, and the base ShopStockModifier.
Something like `public abstract class BarterShopStockModifier: ShopStockModifier`
Then it can hold these methods, and both of the child classes can use them trivially